src/lib/auth/redux/authSlice.js

```diff
@@ -0,0 +1,231 @@
+import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
+import { buildUrlEncodedData } from "../utils";
+import { LS_BENTO_WAS_SIGNED_IN, setLSNotSignedIn } from "../performAuth";
+import { decodeJwt } from "jose";
+import { makeResourceKey } from "../resources";
+
+export const tokenHandoff = createAsyncThunk(
+    "auth/TOKEN_HANDOFF",
+    async ({ code, verifier, clientId, authCallbackUrl }, { getState }) => {
+        const url = getState().openIdConfiguration.data?.["token_endpoint"];
+
+        const response = await fetch(url, {
+            method: "POST",
+            headers: { "Content-Type": "application/x-www-form-urlencoded" },
+            body: buildUrlEncodedData({
+                grant_type: "authorization_code",
+                code,
+                client_id: clientId,
+                redirect_uri: authCallbackUrl,
+                code_verifier: verifier,
+            }),
+        });
+
+        // Assuming the server responds with JSON
+        return await response.json();
+    }
+);
+
+export const refreshTokens = createAsyncThunk("auth/REFRESH_TOKENS", async ({ clientId }, { getState }) => {
+    if (!getState().auth.refreshToken) return;
+
+    const url = getState().openIdConfiguration.data?.["token_endpoint"];
+
+    const response = await fetch(url, {
+        method: "POST",
+        headers: { "Content-Type": "application/x-www-form-urlencoded" },
+        body: buildUrlEncodedData({
+            grant_type: "refresh_token",
+            client_id: clientId,
+            refresh_token: getState().auth.refreshToken,
+        }),
+    });
+
+    // Assuming the server responds with JSON
+    return await response.json();
+});
+
+export const fetchResourcePermissions = createAsyncThunk(
+    "auth/FETCH_RESOURCE_PERMISSIONS",
+    async ({ resource }, thunkAPI) => {
+        if (!thunkAPI.getState().services.itemsByKind.authorization?.url) {
+            console.error("Missing authorization service");
+            return;
+        }
+        const key = makeResourceKey(resource);
+        const rp = thunkAPI.getState().auth.resourcePermissions[key];
+        if (rp?.isFetching || rp?.permissions) {
+            return;
+        }
+        // TODO: ask for it
+        const url = `${thunkAPI.getState().services.itemsByKind.authorization.url}/policy/permissions`;
+        const response = await fetch(url, {
+            method: "POST",
+            headers: { "Content-Type": "application/json" },
+            body: JSON.stringify({ requested_resource: resource }),
```

Comment:
this system has changed a bit, see latest changes to bento web - we'll need to incorporate that

Reply:
this isn't resolved, not sure why it was marked as resolved

Reply:
welp, now it is